### ISSUE TITLE
Unable to configure aiomonitor to listen on the 0.0.0.0 address

### ISSUE THREAD

**Issue by scottctaylor12** [mitre/caldera#111]
By default, the application has ports 8880 and 8888 listen on the loopback interface, 127.0.0.1. When I run `netstat -tulpn` I can confirm this:

tcp        0      0 127.0.0.1:8880          0.0.0.0:*               LISTEN      3893/python3
tcp        0      0 127.0.0.1:8888            0.0.0.0:*               LISTEN      3893/python3

However, I would like configure both ports to listen on the 0.0.0.0 address to accept all remote connections. To do this, I go into the `conf/local.yml` file and change the host from 127.0.0.1 to 0.0.0.0. After restarting the application with the new configurations, I see the startup logs state:

INFO:aiomonitor.monitor:Starting aiomonitor at 127.0.0.1:8880
DEBUG:root:Starting CALDERA at 0.0.0.0:8888

The web interface works great and I can navigate to it from anywhere in my network. However, the API accessed through netcat is still only listening on the loopback address. A `netstat -tulpn` command output shows:

tcp        0      0 **127.0.0.1**:8880          0.0.0.0:*               LISTEN      3930/python3
tcp        0      0 0.0.0.0:8888            0.0.0.0:*               LISTEN      3930/python3

TLDR; the API on port 8880 needs to change what interface it listens on when the conf/local.yml file is edited with a new host address.

**Comment by privateducky** [comment]
This is an interesting topic. When you swap from loopback to all interfaces, we intentionally keep the terminal app on the loopback (as you’ve noticed). 

The reason is authentication: the web (8888) app is protected by login/logout credentials whereas the terminal app is not. This interface is intended for local usage only, as it allows lower-level access. 

I thought about wrapping the terminal app with the same auth as the web app, but figured it was just as easy (and safer) to access the box remotely and then pop into the terminal. Sort of like how msfconsole is used on Metasploit.

**Comment by scottctaylor12** [comment]
That completely makes sense assuming that this is running on a standard Linux system. However, the reason I bring up this issue is because I am attempting to containerize CALDERA. I have my progress posted over at https://github.com/scottctaylor12/caldera-docker and I am able to get the web interface to work totally fine. Because I am using a basic python container, tools like netcat are not included in the image... Also it is not natural to manually log into a running container to access the API.
I completely understand the reasoning for protecting the API due to the lack of authentication. I am just in a situation where docker is not playing nicely. What I have to do is run `docker container exec <container name> bash` to get an interactive session, `apt-get install netcat`, and then `netcat localhost 8880` to access the API. Let me know what your thoughts are given my docker situation.
My long term goal is to setup an entire attacking infrastructure with a single docker-compose.yml file and I think CALDERA would be a great component in that environment :)

**Comment by khyberspache** [comment]
@scottctaylor12 You don't *have* to use NetCat to console in.  Realistically all you need is a raw TCP socket connection, so you have a couple options:
- Try TelNet, might work?
- Python2.7 based script (import socket, bind to localhost:8880, and write minimal console handler).  This is probably the easiest, best cross-platform approach.
- Depending on linux distro, use the /dev/tcp functionality (https://laconicwolf.com/2017/08/13/netcat-alternative-devtcp/)
- You could also try some trickery with TMUX/Screen by attaching to the localhost socket file (absolutely no idea if this is possible)...

**Comment by privateducky** [comment]
another potential option, is when the terminal app is initialized (server.py, line 67) it should be possible to specify the host address of the terminal. by default, it'll use 127.0.0.1. i'd be open to a pull request that adds a new property to local.yml (terminal_host: 0.0.0.0) so when booting up the server, you can decide the host for the web app & the terminal app, separately.

**Comment by scottctaylor12** [comment]
@khyberspache @privateducky thanks for the feedback and potential options! I quickly tested the /dev/tcp idea and unfortunately I did not get that one to connect properly. In order to use screen, tmux, or a telnet client, I would have to install it in the container which I am trying to avoid. 

Tonight I will try passing a python script into the container so I can do something like `docker container exec caldera 'python api-connect.py'`. If that fails, then I will try modifying the server.py file to accept a terminal_host variable from the local.yml. Considering this is in a container, I could always configure to host-based firewall to control external port 8880 connections so it is not _completely_ open to the world unauthenticated!

For reference, my docker setup is documented at https://github.com/scottctaylor12/caldera-docker

**Comment by privateducky** [comment]
@scottctaylor12 sounds great - let me know how it goes. thanks for sharing the docker-compose-- we've been wanting to drop in a Docker image at the root of the core CALDERA code-- so if you're open to it, we'd love to pull in your final copy. keep me posted.

**Comment by scottctaylor12** [comment]
I have added a `terminal_host` variable in the conf/local.yml file and edited server.py to interpret it. Now one can change the interface that the API binds to. 

On top of that, I have finalized the docker-compose.yml  file and it works if the `host` and `terminal_host` variable is set to 0.0.0.0 or the host machine's IP. All someone has to do is, from the root of the CALDERA directory, run `docker-compose up -d` and the application will start in docker!

All of this can be found in pull request #116 :)

**Comment by privateducky** [comment]
merged the final solution. thanks @scottctaylor12!

**Comment by scottctaylor12** [comment]
Awesome! Hopefully the community enjoys the docker feature. Now, I'll start looking into writing some plugin.
